source/common/network/io_uring_socket_handle_impl.cc

```diff
@@ -0,0 +1,419 @@
+}
+
+Api::IoCallUint64Result
+IoUringSocketHandleImpl::readv(uint64_t max_length, Buffer::RawSlice* slices, uint64_t num_slice) {
+  ENVOY_LOG(debug, "readv, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+
+  Api::IoCallUint64Result result = copyOut(max_length, slices, num_slice);
+  if (result.ok()) {
+    // If the return value is 0, there should be a remote close. Return the value directly.
+    if (result.return_value_ != 0) {
+      io_uring_socket_->getReadParam()->buf_.drain(result.return_value_);
+    }
+  }
+  return result;
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::read(Buffer::Instance& buffer,
+                                                      absl::optional<uint64_t> max_length_opt) {
+  ENVOY_LOG(trace, "read, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+
+  absl::optional<Api::IoCallUint64Result> read_result = checkReadResult();
+  if (read_result.has_value()) {
+    return std::move(*read_result);
+  }
+
+  const OptRef<Io::ReadParam>& read_param = io_uring_socket_->getReadParam();
+  uint64_t max_read_length =
+      std::min(max_length_opt.value_or(UINT64_MAX), read_param->buf_.length());
+  buffer.move(read_param->buf_, max_read_length);
+  return {max_read_length, IoSocketError::none()};
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::writev(const Buffer::RawSlice* slices,
+                                                        uint64_t num_slice) {
+  ENVOY_LOG(trace, "writev, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+
+  absl::optional<Api::IoCallUint64Result> write_result = checkWriteResult();
+  if (write_result.has_value()) {
+    return std::move(*write_result);
+  }
+
+  uint64_t ret = io_uring_socket_->write(slices, num_slice);
+  return {ret, IoSocketError::none()};
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::write(Buffer::Instance& buffer) {
+  ENVOY_LOG(trace, "write {}, fd = {}, type = {}", buffer.length(), fd_, ioUringSocketTypeStr());
+
+  absl::optional<Api::IoCallUint64Result> write_result = checkWriteResult();
+  if (write_result.has_value()) {
+    return std::move(*write_result);
+  }
+
+  uint64_t buffer_size = buffer.length();
+  io_uring_socket_->write(buffer);
+  return {buffer_size, IoSocketError::none()};
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::sendmsg(const Buffer::RawSlice*, uint64_t, int,
+                                                         const Address::Ip*,
+                                                         const Address::Instance&) {
+  ENVOY_LOG(trace, "sendmsg, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+  return Network::IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::recvmsg(Buffer::RawSlice*, const uint64_t,
+                                                         uint32_t, RecvMsgOutput&) {
+  ENVOY_LOG(trace, "recvmsg, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+  return Network::IoSocketError::ioResultSocketInvalidAddress();
```

Comment:
For my education, are sendmsg/recvmsg not supported for io_uring?

Reply:
Both sendmsg and recvmsg are supported by io_uring. However, at this stage we only implement TCP-related methods, and we will also determine whether the socket is TCP or not to decide whether to enable io_uring in the socket interface patch later.